### PR TITLE
couple o` bugfixes

### PR DESCRIPTION
**Pull request by thet**
from the commit log:
- simplify makefile: all, clean, install, serve. no stamps, no designerhappy
- currently we need the resolutions section for jquery
- why not using latest http-server?
- removed stamp-bower and stamp-npm from Makefile, now also from gitignore
- add pattern class to dom element for a real demo. note NEEDS data element, otherwise parser breaks? reformat using 2 spaces indentation
- fix mainfile: let require depend on our app, wait for document loaded until running registry.init(), add missing dependencies, remove unnecessary dependencies
- let pattern do something simple for demo effects. name logger after app name. log something.
- require app_name, create app_name directory instead using current one, write main.js template, write indexfile with template substitions replaced correctly, restructure
- raise version

to test it, check this out, cd into it and do:

```
$ npm install
$ npm link
$ yo patternslib pat-test
```

### CODE REVIEW

**Comment by thet** [comment]
@jcbrand 
there is one thing, which confuses me: mockup-parser is loaded by require.js from /src/bower_components/patternslib/src/core/mockup-parser.js
IMO, it shouldn't.
do i have to explicitly state the parser to use? does it default to the mockup one?
i encountered, mockup-parser throwing an error, if the `data-APP_NAME` is missing on the dom element, which gets initialized. actually, this is another story, but i'm stating it here.

File: app/index.js [hunk 1]
```diff
@@ -7,58 +7,50 @@ var PatternGenerator = yeoman.generators.Base.extend({
...
     yeoman.generators.Base.apply(this, arguments);
 
+    this.argument('app_name', { type: String, required: true });
+    this.appname = this.app_name || this.appname;
     this.option('appPath', {
       desc: 'Name of application directory',
       type: 'String',
-      defaults: './',
+      defaults: this.appname,
       banner: 'some banner'
```
Comment (jcbrand) [inline app/index.js:16]: Nice that you fixed this.

File: app/index.js [hunk 2]
```diff
@@ -7,58 +7,50 @@ var PatternGenerator = yeoman.generators.Base.extend({
...
+    this.env.options.appPath = this.options.appPath;
     this.config.set('appPath', this.env.options.appPath);
-    this.indexFile = this.readFileAsString(path.join(this.sourceRoot(), 'index.html'));
   },
 
   writing: {
-    setupEnv: function () {
+
+    setup_base: function () {
       this.mkdir(this.env.options.appPath);
```
Comment (jcbrand) [inline app/index.js:25]: By convention, Javascript methods are written in bumpyCase, so should be `setupBase`, same for the others.

**Comment by jcbrand** [comment]
> simplify makefile: all, clean, install, serve. no stamps, no designerhappy

`designerhappy` is a convention for Cornelis.

> currently we need the resolutions section for jquery

Makes sense.

> why not using latest http-server?

No specific reason. Probably it was released after the generator was created?

>   removed stamp-bower and stamp-npm from Makefile, now also from gitignore

Why? We use them in Patternslib, in Plone Intranet and many other projects. It's a convention used to avoid unnecessary time spent with bower and NPM updates when files are already up to date.

>  add pattern class to dom element for a real demo. note NEEDS data element, otherwise parser breaks?

No, should definitely not need a data element. Can you give an example/traceback of the parser breaking? (note: which parser are you referring to here?)

Also, we don't know what DOM element the pattern applies to, so just applying it to any DOM element doesn't sound particularly helpful to me.

> reformat using 2 spaces indentation

Patternslib uses 4 spaces for indentation. All of the patternslib code follows this.
I know Mockup uses only 2. Just another one of the many areas in which Mockup does things differently.

I really don't like the 2 space indentation (it makes the code look squashed), and although many JS projects use 2 spaces, many instead use 4 spaces.

>  fix mainfile: let require depend on our app, wait for document loaded until running registry.init(), add missing dependencies, remove unnecessary dependencies
>  let pattern do something simple for demo effects. name logger after app name. log something.
> require app_name, create app_name directory instead using current one, write main.js template, write indexfile with template substitions replaced correctly, restructure
>  raise version

Last four points all sound good.

**Comment by jcbrand** [comment]
> there is one thing, which confuses me: mockup-parser is loaded by require.js from /src/bower_components/patternslib/src/core/mockup-parser.js
> IMO, it shouldn't.

Why not? What do you expect instead?

> do i have to explicitly state the parser to use? does it default to the mockup one?

It defaults to the Patternslib parser and if you want to use the Mockup parser (which we should all stop using) you need to specify `'parser': 'mockup'`.

> i encountered, mockup-parser throwing an error, if the data-APP_NAME is missing on the dom element, which gets initialized. actually, this is another story, but i'm stating it here.

The mockup parser also doesn't properly parse Booleans. `false` gets turned into `'false'` which is a truthy value, causing patterns to break when they expect booleans to be handled properly.

File: app/index.js (hunk 2 above)
Comment (thet) [inline app/index.js:25]: Patternslib style guide? I prefer pep8 style more, because of better visual seperation.
However, will fix.

Comment (jcbrand) [inline app/index.js:25]: Not really Patternslib, this is a very general Javascript convention. Comes from Douglas Crockfords' "Javascript the good parts" and also other areas.

**Comment by thet** [comment]
> designerhappy is a convention for Cornelis.

i thought so, but removed because `all` is shorter and clearer. if we need Cornelis Convention Compatiblity, i'll add it back.

> Why? We use them in Patternslib, in Plone Intranet and many other projects. It's a convention used to avoid unnecessary time spent with bower and NPM updates when files are already up to date.

found no signs of use within the generated files itself. so, patternslib is checking for stamp-bower/-npm timestamps? is patternslib also making use of this information for external packages like those generated by patternslib? would make sense. would add back then.

> Also, we don't know what DOM element the pattern applies to, so just applying it to any DOM element doesn't sound particularly helpful to me.

i think you misunderstood: let's say we create a pattern with the name `pat-test`. before there was no `pat-test` class in the HTML anywhere, now there is. also now there is a `data-pat-test`, which is apparently needed by the mockup parser (will re-check).

> I really don't like the 2 space indentation (it makes the code look squashed), and although many JS projects use 2 spaces, many instead use 4 spaces.

is there a coding convention document? if not, we have to define conventions and write them down. then i'd propose to use it for Plone's JS coding conventions too.

personally i like 2 spaces for XML, because XML+HTML is so lengthy and often has a deep hierarchy, that it saves a lot of screen estate. also Zope's ZCML styleguide recommends 2 spaces and makes a lot of sense for me: http://docs.zope.org/zopetoolkit/codingstyle/zcml-style.html#tabs-and-spaces 

i'm not sure about 2 spaces in javascript files, though. i'm used to use 2 spaces, but i agree, the code looks quite squashed.

however, i'll fix to meet patternslib conventions.

**Comment by jcbrand** [comment]
> is there a coding convention document? if not, we have to define conventions and write them down.
- The coding styleguide is here: http://patternslib.com/pattern-styleguide/#main-content
- The .jshint file in the Patternslib directory also encodes many of the coding conventions.
- I also made a styleguide (in Sphinx) for converse.js which has some overlap with the style in Patternslib: https://conversejs.org/docs/html/style_guide.html

The styleguide on Patternslib.com should definitely be expanded upon, currently too many conventions are implicit and not explicitly mentioned.

> then i'd propose to use it for Plone's JS coding conventions too.

I'd be very happy if we could find a common convention for both projects.

I actually also want to still write a blog post also about the use of `var self = this` in Mockup, which IMO has certain drawbacks and is not idiomatic Javascript.

> personally i like 2 spaces for XML, because XML+HTML is so lengthy and often has a deep hierarchy, that it saves a lot of screen estate. also Zope's ZCML styleguide recommends 2 spaces and makes a lot of sense for me: http://docs.zope.org/zopetoolkit/codingstyle/zcml-style.html#tabs-and-spaces

Yeah, makes sense to me as well.

> i'm not sure about 2 spaces in javascript files, though. i'm used to use 2 spaces, but i agree, the code looks quite squashed.
> 
> however, i'll fix to meet patternslib conventions.

Thanks!

**Comment by thet** [comment]
> No, should definitely not need a data element.

i re-checked, and it doesn't need that element. was a follow-up error then.

> It defaults to the Patternslib parser and if you want to use the Mockup parser (which we should all stop using) you need to specify 'parser': 'mockup'.

that's what i expected. but looking at the network tab of the browser inspector or the console output of the `make serve` command, the `mockup-parser.js` is loaded. might not be a problem, but feels unnecessary for me.

so, 2 spaces for html or just 4 for everything?

**Comment by jcbrand** [comment]
> that's what i expected. but looking at the network tab of the browser inspector or the console output of the make serve command, the mockup-parser.js is loaded. might not be a problem, but feels unnecessary for me.

Yes, sounds unnecessary. I think Patternslib pulls it in as a core dependency. Howeve, the mockup parser is tiny, it's 50 lines in total, so not a big deal when in a bundle and minified. However, if you have the time, you're welcome to debug why it gets pulled in and whether it's necessary.

> so, 2 spaces for html or just 4 for everything?

You can decide. I often do HTML with 4 spaces, but there are lots of different ways in which people write HTML. I don't have a strong preference there.

**Pull request closed without merging**